compiler/src/formatting/fmt.re

```diff
@@ -0,0 +1,3432 @@
+        };
+      };
+
+      // We use phantom hardlines here to prevent comment line breaks from
+      // disrupting the formatting engine's width calculations. This means that
+      // the engine guarantees we get a break for line comments, but it
+      // considers the following code as just one really long line:
+      //     func(foo, // comment
+      //       bar, baz, // comment
+      //       qux // comment
+      //     )
+      // This gives us a balance between comments completely changing the
+      // formatting of an expression versus the engine just being "mindful"
+      // about the comments.
+
+      switch (comments) {
+      | [] => empty
+      | comments =>
+        concat_map(
+          ~lead=
+            next =>
+              switch (loc(next).loc_start.pos_lnum - prev.loc_end.pos_lnum) {
+              | _ when block_start => lead
+              | 0 => lead
+              | 1 when allow_breaks => phantom_hardline
+              | _ when allow_breaks => phantom_hardline ++ phantom_hardline
+              | _ => lead
+              },
+          ~sep=
+            (prev, next) => {
+              let line_delta =
+                loc(next).loc_start.pos_lnum - loc(prev).loc_end.pos_lnum;
+              if (allow_breaks) {
+                switch (prev) {
+                | Block(_) when line_delta == 0 => breakable_space
+                | _ when line_delta <= 1 => phantom_hardline
+                | _ => phantom_hardline ++ phantom_hardline
+                };
+              } else {
+                switch (prev) {
+                | Line(_)
+                | Shebang(_) => phantom_hardline // required for line comments no matter what
+                | _ => space
+                };
+              };
+            },
+          ~trail=
+            prev => {
+              let line_delta =
+                next.loc_start.pos_lnum - loc(prev).loc_end.pos_lnum;
+              switch (prev) {
+              // prevent double spacing of comments at the end of a block
+              | Line(_)
+              | Shebang(_) when block_end =>
+                if_broken(empty, phantom_hardline)
+              | _ when block_end => empty
+              | Line(_)
+              | Shebang(_) when line_delta <= 1 => phantom_hardline
+              | Line(_)
+              | Shebang(_) when line_delta > 1 =>
+                phantom_hardline ++ phantom_hardline
+              | Doc(_) when allow_breaks => phantom_hardline
+              | _ when allow_breaks && line_delta == 1 => phantom_hardline
+              | _ when allow_breaks && line_delta > 1 =>
+                phantom_hardline ++ phantom_hardline
+              | _ => trail
+              };
+            },
+          ~f=(~final, c) => print_comment(c),
+          comments,
+        )
+      };
+    };
+  };
+
+  let toplevel =
+    switch (parsed_program.statements) {
+    | [] =>
+      print_comment_range(
+        ~none=hardline,
+        ~lead=space,
```

Comment:
What is this supposed to do? I removed it and no tests changed.

Reply:
I actually removed `~trail` and `~none` too and nothing changed.

Reply:
There just wasn't a test. Test added.